flutter-idea/src/io/flutter/view/EmbeddedBrowser.java

```diff
@@ -0,0 +1,185 @@
+import com.teamdev.jxbrowser.browser.callback.ConfirmCallback;
+import com.teamdev.jxbrowser.browser.callback.input.PressKeyCallback;
+import com.teamdev.jxbrowser.ui.KeyCode;
+import com.teamdev.jxbrowser.ui.event.KeyPressed;
+import com.teamdev.jxbrowser.view.swing.BrowserView;
+import com.teamdev.jxbrowser.view.swing.callback.DefaultAlertCallback;
+import com.teamdev.jxbrowser.view.swing.callback.DefaultConfirmCallback;
+import icons.FlutterIcons;
+import io.flutter.FlutterInitializer;
+import io.flutter.devtools.DevToolsUrl;
+import io.flutter.jxbrowser.EmbeddedJxBrowser;
+import io.flutter.utils.AsyncUtils;
+import org.jetbrains.annotations.NotNull;
+
+import javax.swing.*;
+import java.awt.*;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+
+class BrowserTab {
+  protected EmbeddedTab embeddedTab;
+  protected Content content;
+  protected CompletableFuture<DevToolsUrl> devToolsUrlFuture;
+}
+
+public abstract class EmbeddedBrowser {
+  protected final Map<@NotNull String, @NotNull BrowserTab> tabs = new HashMap<>();
+
+  public abstract Logger logger();
+
+  public EmbeddedBrowser(Project project) {
+    ProjectManager.getInstance().addProjectManagerListener(project, new ProjectManagerListener() {
+      @Override
+      public void projectClosing(@NotNull Project project) {
+        for (final String tabName: tabs.keySet()) {
+          final BrowserTab tab = tabs.get(tabName);
+          if (tab.embeddedTab != null) {
+            try {
+              tab.embeddedTab.close();
+            } catch (Exception ex) {
+              logger().info(ex);
+            }
+            tabs.remove(tabName);
+          }
+        }
+      }
+    });
+  }
+
+  public void openPanel(ContentManager contentManager, String tabName, DevToolsUrl devToolsUrl, Runnable onBrowserUnavailable) {
+    final BrowserTab firstTab = tabs.get(tabName);
+    if (firstTab == null) {
+      openBrowserTabFor(tabName);
+    }
+    final BrowserTab tab = tabs.get(tabName);
+    // If the browser failed to start during setup, run unavailable callback.
+    if (tab == null) {
+      onBrowserUnavailable.run();
+      return;
+    }
+
+    // Multiple LoadFinished events can occur, but we only need to add content the first time.
+    final AtomicBoolean contentLoaded = new AtomicBoolean(false);
+
+    try {
+      tab.embeddedTab.loadUrl(devToolsUrl.getUrlString());
+    } catch (Exception ex) {
+      tab.devToolsUrlFuture.completeExceptionally(ex);
+      onBrowserUnavailable.run();
+      logger().info(ex);
+      return;
+    }
+
+    tab.devToolsUrlFuture.complete(devToolsUrl);
+
+    JComponent component = tab.embeddedTab.getTabComponent(contentManager);
+
+    ApplicationManager.getApplication().invokeLater(() -> {
+      if (contentManager.isDisposed()) {
+        return;
+      }
+
+      contentManager.removeAllContents(false);
```

Comment:
I'm not sure if `setSelectedContent` to `previousContent` from this previous PR needs to be implemented here as well: https://github.com/flutter/flutter-intellij/pull/6772/files#diff-7475425770d21bb97a1d3a0796d3cd02868371cde331681a8812191064a1f6d2R165

What are the implications of not setting selected to previous?

Reply:
There is nothing specific here, I wrote this mainly to avoid if `Content`s have different properties between the same tab, so I was thinking it would be safer to remove the previous one then open a new one.

Reply:
Okay, sounds like this should be okay then. Thanks!